### PR TITLE
fix(engine): Partially fix non-resolvable-interface-object

### PR DESCRIPTION
**Pull request by Finistere**
So the primary goal of this PR was to solve the `non-resolvable-interface-object` test, which it improves. The problem was that we query:
```graphql
query {
  node {
     id
  }
}
```

with a schema like:

```graphql
type Query {
  node: Node!
}

interface Node {
  id: ID!
}
```

Not that `Node` has no implementors but the subgraph still returns data. However, with the current design when I needed to have at least one implementor.

Let's take a step back, to deserialize data I use extensively `DeserializeSeed` to avoid any extra allocations like `serde_json::Value` would. But this means that I need to know what kind of data I expect during deserialization. To do this we construct all the possible field and object shapes we may encounter. This is potentially an extremely costly step so I had already made some effort in an efficient partitioning logic to partition the possible output types into distinct partitions that each have the same set of fields and thus the same concrete shape. If we have:
```graphql
interface Node {
  id: ID!
}

interface Other {
  other: String!
}

type A implements Node & Other { ... }
type B implements Node { .. }

type Query {
  node: Node!
}
```

then `{ node { id ... on Other { other } } }` has two distinct partitions: `[[A], [B]]` as they each have different fields. but if we only query `{ node { id } }` then we have a single partition `[A, B]`.

The current logic didn't keep track of the associated fields with each partition though. So I used any object id in the partition and iterated over all fields filtering out the ones that don't apply. And here we come back to the `non-resolvable-interface-object` test, as there wasn't any implementors, we didn't have any object and couldn't create the field shape!

So to fix that and improve the performance, as field filtering was slow, I'm now keeping track of fields in a bitset. This significantly improved performance and for fun I removed the other significant performance problem which was keeping track of which field shape was associated with which field. Now I'm avoid most allocations for this and using something similar to a counting sort. Both of those reduce by a 10x factor the time it takes for operation preparation in my worst-case scenario (440ms -> ~40ms) and 3x for the simplest bench I have for this (1ms -> 300us). So small extra win.

While the partitioning logic is tested with unit-tests I didn't have any decent tests at the GraphQL layer including how shapes are used and not only generated. So wrote a bunch of those. And hit a corner case with fragments...
For a schema like:
```graphql
interface Node {
  id: ID!
}

interface Other {
  other: String!
}

type A implements Node & Other { ... }
type B implements Node { .. }
type C implements Other { .. }

type Query {
  node: Node!
}
```
It is valid to write:
```graphql
query {
  node {
    ...Frag
  }
}

fragment Frag on Other {
   ... on C { other }
}
```

But this ends up as: `query { node { ... on C { other } } }` in the engine as we flatten all type conditions. We weren't removing those fields and instead were requesting those in subgraph queries... So did a quick hack to solve it, filtering them out in the binding step. But will move it to the planning step and simplify the binding step to keep fragments later.

### CODE REVIEW

**Review by tomhoule — APPROVED** [review]
It's a lot to absorb, I can't say I get every detail, but the changes in the partitioning / shape generation are great, seems like a simplifying step to keep track of the fields in the bitset.

File: crates/engine/src/operation/solve/solver/shapes/mod.rs [hunk 1]
```diff
@@ -22,352 +22,438 @@ use super::Solver;
...
         typename_fields: &[TypenameField<'ctx>],
-    ) -> Option<Vec<Partition<ObjectDefinitionId>>> {
-        let mut type_conditions = Vec::new();
-        for field in typename_fields {
-            type_conditions.push(field.type_condition_id);
+        data_fields: &[DataField<'ctx>],
+    ) -> partition::Partitioning<ObjectDefinitionId, FixedBitSet> {
+        let mut type_condition_and_field_position_in_bitset =
+            Vec::with_capacity(typename_fields.len() + data_fields.len());
+        for (i, field) in typename_fields.iter().enumerate() {
```
Comment (pimeys) [inline crates/engine/src/operation/solve/solver/shapes/mod.rs:429]: Here the `i` is fine due to the enumerate :)

File: crates/engine/src/operation/solve/solver/shapes/partition.rs [hunk 2]
```diff
@@ -37,287 +39,233 @@ impl<Id: Copy> From<Vec<Id>> for Partition<Id> {
...
+    type_condition_possible_types: &[Id],
+    output_possible_types: &[Id],
+) -> bool {
+    if output_possible_types.len() > type_condition_possible_types.len() {
         return false;
     }
-    let mut i = 0;
-    let mut j = 0;
+    let mut l = 0;
+    let mut r = 0;
```
Comment (pimeys) [inline crates/engine/src/operation/solve/solver/shapes/partition.rs:139]: What are these variables? Either comment or rename.

File: crates/engine/src/operation/solve/solver/shapes/partition.rs [hunk 3]
```diff
@@ -37,287 +39,233 @@ impl<Id: Copy> From<Vec<Id>> for Partition<Id> {
...
+        new_partition.clear();
 
-        // remaining = (possible_types ∩ selection_set_possible_types)
-        let mut i = 0;
-        let mut j = 0;
-        while let Some((left, right)) = possible_types.get(i).zip(selection_set_possible_types.get(j)) {
+        // Initialize the new partition to the intersection with the output possible types
+        // new_partition = (possible_types ∩ output_possible_types)
+        let mut l = 0;
+        let mut r = 0;
```
Comment (pimeys) [inline crates/engine/src/operation/solve/solver/shapes/partition.rs:179]: Same here.

File: crates/engine/src/operation/solve/solver/shapes/partition.rs [hunk 4]
```diff
@@ -37,287 +39,233 @@ impl<Id: Copy> From<Vec<Id>> for Partition<Id> {
...
 
 // Both arrays are sorted an non-empty.
-fn extract_intersection<Id: Copy + Ord>(existing: &mut Vec<Id>, candidate: &mut Vec<Id>, intersection: &mut Vec<Id>) {
-    let mut i = 0;
-    let mut j = 0;
-    let mut new_existing_len = 0;
-    let mut new_candidate_len = 0;
+fn extract_intersection<Id: Copy + Ord>(left: &mut Vec<Id>, right: &mut Vec<Id>, intersection: &mut Vec<Id>) {
+    let mut l = 0;
+    let mut r = 0;
```
Comment (pimeys) [inline crates/engine/src/operation/solve/solver/shapes/partition.rs:223]: And here.

File: crates/engine/src/operation/solve/solver/shapes/mod.rs [hunk 5]
```diff
@@ -22,352 +22,438 @@ use super::Solver;
...
-            .map(|field| field.key)
-            .collect();
-
-        let mut fields_buffer = self.data_fields_buffer_pool.pop();
         let mut field_shapes_buffer = self.field_shapes_buffer_pool.pop();
+        let mut distinct_typename_response_keys: Vec<PositionedResponseKey> = Vec::new();
+        let mut included = included_typename_then_data_fields.into_ones();
+
+        let mut all_expected_keys_equal_response_keys = true;
+        while let Some(i) = included.next() {
```
[the remark this answers is not in the document]
Comment (Finistere) [reply crates/engine/src/operation/solve/solver/shapes/mod.rs:164]: It's an index, so at best I can do `index` or `ix` if you really think that's better. Otherwise I can add a comment if it's not clear what's going on for you

File: crates/engine/schema/src/composite_type.rs [hunk 6]
```diff
@@ -77,4 +77,28 @@ impl<'a> CompositeType<'a> {
...
+            CompositeType::Interface(interface) => &interface.possible_type_ids,
+            CompositeType::Union(union) => &union.possible_type_ids,
+        }
+    }
+
+    pub fn has_non_empty_intersection_with(&self, other: CompositeType<'a>) -> bool {
+        let left = self.possible_types();
+        let right = other.possible_types();
+        let mut l = 0;
+        let mut r = 0;
```
[the remark this answers is not in the document]
Comment (Finistere) [reply crates/engine/schema/src/composite_type.rs:93]: they're indices, `l` is an index in `left` and `r` an index in `right`. What's a better name in a 10 lines function for you?

[the remark this answers is not in the document]
Comment (Finistere) [reply crates/engine/schema/src/composite_type.rs:93]: It's the same thing everywhere else. I just find `l` for `left` and `r` for `right` clearer than `i` for `left` and `j` for `right`. And for indices, one letter variables are common use and good enough IMHO. So not sure for what you'd like to change them. Any suggestions? Or is the code unclear?

[the remark this answers is not in the document]
Comment (pimeys) [reply crates/engine/schema/src/composite_type.rs:93]: You can merge this if you think it's clear. I'm just having no idea what these variables are, so having a name describing exactly what index we are talking about would help...

File: crates/graphql-mocks/src/dynamic/builder.rs [hunk 7]
```diff
@@ -136,8 +139,37 @@ fn convert_object(def: parser::ObjectDefinition<'_>, resolvers: &mut ResolverMap
     object.into()
 }
 
-fn convert_iface(_def: parser::InterfaceDefinition<'_>) -> async_graphql::dynamic::Type {
-    todo!("implement interfaces if you want, probably just copy the object impl")
+fn transform_into_field_value(mut value: async_graphql::Value) -> FieldValue<'static> {
+    match value {
+        async_graphql::Value::Object(ref mut fields) => {
+            if let Some(async_graphql::Value::String(ty)) = fields.swap_remove("__typename") {
```
[the remark this answers is not in the document]
Comment (Finistere) [reply crates/graphql-mocks/src/dynamic/builder.rs:145]: yeah did naturally with `__typename` and then found out it obviously didn't just work xD So made it, luckily wasn't too hard, just had to understand what to do

File: crates/engine/schema/src/composite_type.rs (hunk 6 above)
[the remark this answers is not in the document]
Comment (tomhoule) [reply crates/engine/schema/src/composite_type.rs:93]: FWIW I parsed it as "left" and "right" without thinking about it.

[the remark this answers is not in the document]
Comment (Finistere) [reply crates/engine/schema/src/composite_type.rs:93]: well that's the intent. Think it makes it a bit clearer that `l` is the index in `left` and `r` the index in `right`, rather than `i` and `j`.
